File: LAS_testing.R
library(tidyverse)
library(pracma)
library(data.table)
library(minpack.lm)
library(plotly)
library(mclust)

tidy_conversion <- function(data) {
   x_axis <- data[, 1]
   data2 <- data |>
       rename(x_axis = V1) |>
       pivot_longer(
           cols = paste0("V", seq(2:length(data)) + 1),
           names_to = "id", 
           values_to = "intensity"
       ) |>
       mutate(id = as.numeric(sub("V", "", id)) - 1)
}

find_peak_locations <- function(data) {
    data <- data[data$V1 > 375 & data$V1 < 420, ]
    x_axis <- data$V1
    
    e2g_idx <- which(x_axis > 375 & x_axis < 395)
    a1g_idx <- which(x_axis > 395 & x_axis < 420)
    results <- vector("list", ncol(data) - 1)
    
    for (i in 2:ncol(data)) {
        intensity <- data[[i]]
        
        peak1 <- e2g_idx[which.max(intensity[e2g_idx])]
        peak2 <- a1g_idx[which.max(intensity[a1g_idx])]
        
        results[[i-1]] <- tibble(
            id = i - 1,
            x_axis1 = x_axis[peak1],
            intensity1 = intensity[peak1],
            x_axis2 = x_axis[peak2],
            intensity2 = intensity[peak2]
        )
        print(paste0("Spectrum ", (i - 1), " processed."))
    }
    peak_locations <- bind_rows(results)
    peak_locations
}

gaussian <- function(x, A, mu, sigma) {
    A * exp(-(x - mu) ^ 2 / (2 * sigma ^ 2))
}

double_gaussian <- function(x, A1, mu1, sigma1, A2, mu2, sigma2, C) {
    gaussian(x, A1, mu1, sigma1) + gaussian(x, A2, mu2, sigma2) + C
}

auto_gaussian_summary <- function(data, peak_locations) {
    data <- data[data$V1 > 375 & data$V1 < 420, ]
    x <- data$V1
    results <- vector("list", nrow(peak_locations))
    
    for (i in seq_len(nrow(peak_locations))) {
        spectrum_id <- peak_locations$id[i]
        y <- data[[spectrum_id + 1]]
        
        A1_guess <- peak_locations$intensity1[i]
        A2_guess <- peak_locations$intensity2[i]
        mu1_guess <- peak_locations$x_axis1[i]
        mu2_guess <- peak_locations$x_axis2[i]
        
        fit <- tryCatch({
            nlsLM(y ~ double_gaussian(x, A1, mu1, sigma1, A2, mu2, sigma2, C),
                  start = list(
                      A1 = A1_guess, mu1 = mu1_guess, sigma1 = 3,
                      A2 = A2_guess, mu2 = mu2_guess, sigma2 = 3,
                      C = min(y)
                  ),
                  lower = c(0, 375, 0.5, 0, 395, 0.5, 0),
                  upper = c(Inf, 395, 15, Inf, 420, 15, Inf)
                  )
        },
        error = function(e) {
            cat("Spectrum:", spectrum_id, "\n", e$message, "\n\n")
            NULL
        })
        
        if (is.null(fit)) {
            results[[i]] <- tibble(
                id = spectrum_id, mu1 = 0, mu2 = 0, fwhm1 = 0, fwhm2 = 0,
                A1 = 0, A2 = 0, area1 = 0, area2 = 0, area_ratio = 0,
                snr = 0, rmse = 0, r_squared = 0, diff = 0
            )
            next
        }
        
        p <- coef(fit)
        fwhm1 <- 2.35482 * p["sigma1"]
        fwhm2 <- 2.35482 * p["sigma2"]
        
        area1 <- p["A1"] * p["sigma1"] * sqrt(2 * pi)
        area2 <- p["A2"] * p["sigma2"] * sqrt(2 * pi)
        area_ratio <- area1 / area2
        
        fitted_y <- double_gaussian(
            x, p["A1"], p["mu1"], p["sigma1"], 
            p["A2"], p["mu2"], p["sigma2"], p["C"]
        )
        residuals <- y - fitted_y
        rmse <- sqrt(mean(residuals^2))
        
        ss_res <- sum(residuals^2)
        ss_tot <- sum((y - mean(y))^2)
        r_squared <- 1 - ss_res / ss_tot
        
        noise <- sd(residuals)
        snr <- ifelse(noise == 0, Inf, max(y) / noise)
        
        results[[i]] <- tibble(
            id = spectrum_id, mu1 = p["mu1"], mu2 = p["mu2"],
            fwhm1 = fwhm1, fwhm2 = fwhm2, A1 = p["A1"], A2 = p["A2"],
            area1 = area1, area2 = area2, area_ratio = area_ratio,
            snr = snr, rmse = rmse, r_squared = r_squared,
            diff = abs(p["mu2"] - p["mu1"])
        )
    }
    bind_rows(results)
}

size <- 300
file_path <- paste0(
    "data/default LAS/", 
    size, "x", size, 
    "/Large Area Scan.csv"
    )

compute_time <- round(0.00588271 * size ^ 2 + 2.21832, 2)
paste0("Time to Compute: ", compute_time %/% 60, ":", (compute_time %% 60))
raw <- fread(file_path, header = FALSE)
data <- raw

peak_summary <- find_peak_locations(data)
gaussian_results <- auto_gaussian_summary(data, peak_summary)

peak_summary <- peak_summary |>
    mutate(
        diff = abs(x_axis1 - x_axis2),
        ratio = intensity1 / intensity2,
        ratio = ifelse(ratio > 1, ratio, 1/ratio)
    ) |>
    merge(gaussian_results, by = "id")
''
heatmap_df <- peak_summary |>
    mutate(
        x = ((id - 1) %% size) + 1,
        y = ((id - 1) %/% size) + 1,
        curve = intensity1 > 730 & intensity2 > 730,
        diff.x = ifelse(curve, diff.x, 15),
        diff.x = ifelse(diff.x > 26 | diff.x < 18, 15, diff.x),
        intensity_ratio = ifelse(curve, ratio, 0.9),
        intensity_ratio = ifelse(intensity_ratio > 1.25, 1.25, intensity_ratio),
        mu1 = ifelse(curve, mu1, 0),
        mu2 = ifelse(curve, mu2, 0),
        fwhm1 = ifelse(curve, fwhm1, 0),
        fwhm2 = ifelse(curve, fwhm2, 0),
        A1 = ifelse(curve, A1, 0),
        A2 = ifelse(curve, A2, 0),
        area1 = ifelse(curve, area1, 0),
        area2 = ifelse(curve, area2, 0),
        area_ratio = ifelse(curve, area_ratio, 0),
        snr = ifelse(curve, snr, 0),
        rmse = ifelse(curve, rmse, 0),
        r_squared = ifelse(curve, r_squared, 0),
        diff.y = ifelse(curve, diff.y, 0)
    ) |>
    select(id, curve, x, y, x_axis1, x_axis2, diff.x, mu1, mu2, diff.y, 
           intensity1, intensity2, intensity_ratio, A1, A2, 
           fwhm1, fwhm2, area1, area2, area_ratio, snr, rmse, r_squared)

### PCA ANALYSIS ###
pca_data <- data[data$V1 > 375 & data$V1 < 420, ]
spectra_matrix <- t(as.matrix(pca_data[ , -1]))
spectra_scaled <- scale(spectra_matrix)
pca <- prcomp(spectra_scaled, center = TRUE, scale. = TRUE)

#plot(cumsum(pca$sdev^2 / sum(pca$sdev^2)), type = "b", 
#     xlab = "PC", ylab = "Cumulative Variance Explained")

pca_scores <- as.data.frame(pca$x[, 1:5])
pca_scores$id <- seq_len(nrow(pca_scores))
heatmap_df <- cbind(heatmap_df, pca_scores[, 1:5])

kmeans_vars <- heatmap_df |>
    select(
        diff.x, diff.y, mu1, mu2, intensity_ratio, A1, A2, fwhm1, fwhm2, 
        area1, area2, area_ratio, snr, rmse, r_squared, PC1, PC2, PC3, PC4, PC5
        ) |>
    scale()

cluster_num <- 4
clustering_results <- kmeans(
    kmeans_vars, 
    centers = cluster_num)

ordering <- order(clustering_results$centers[, 1])
mapping <- setNames(1:cluster_num, ordering)
clustering_results$cluster <- mapping[as.character(clustering_results$cluster)]

heatmap_df$cluster <- (clustering_results$cluster - 1) * 6.15

p <- ggplot(heatmap_df, aes(x = x, y = y, fill = cluster)) +
    geom_tile() +
    coord_equal() +
    scale_y_reverse() +
    scale_fill_gradientn(colors = c("lightblue", "yellow", "red")) + 
    labs(fill = "Clustering") + 
    theme_bw()
p
ggplotly(p)

write.csv(heatmap_df, file = "../paraview_data/analysis_results.csv", row.names = FALSE)

# ggplotly(ggplot(data, aes(x = V1, y = V41)) + geom_line() + theme_bw())

File: LAS_testing.py
import os
import numpy as np
import pandas as pd

from scipy.optimize import curve_fit

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

import plotly.express as px

def tidy_conversion(data):
    data2 = data.rename(columns = {0: "x_axis"})

    long_df = data2.melt(
        id_vars = "x_axis",
        var_name = "id",
        value_name = "intensity"
    )

    long_df["id"] = long_df["id"].astype(int)

    return long_df

def find_peak_locations(data):
    data = data[(data[0] > 375) & (data[0] < 420)]
    x_axis = data[0].values

    e2g_idx = np.where((x_axis > 375) & (x_axis < 395))[0]
    a1g_idx = np.where((x_axis > 395) & (x_axis < 420))[0]
    results = []

    for i in range(1, data.shape[1]):
        intensity = data.iloc[:, i].values

        peak1 = e2g_idx[np.argmax(intensity[e2g_idx])]
        peak2 = a1g_idx[np.argmax(intensity[a1g_idx])]

        results.append({
            "id": i,
            "x_axis1": x_axis[peak1],
            "intensity1": intensity[peak1],
            "x_axis2": x_axis[peak2],
            "intensity2": intensity[peak2]
        })
        print(f"Spectrum {i} processed.")

    return pd.DataFrame(results)

def gaussian(x, A, mu, sigma):
    return A * np.exp(-(x - mu)**2 / (2 * sigma**2))

def double_gaussian(x, A1, mu1, sigma1, A2, mu2, sigma2, C):
    return (gaussian(x, A1, mu1, sigma1) + gaussian(x, A2, mu2, sigma2) + C)

def auto_gaussian_summary(data, peak_locations):
    data = data[(data[0] > 375) & (data[0] < 420)]
    x = data[0].values
    results = []

    for _, row in peak_locations.iterrows():
        spectrum_id = int(row["id"])
        y = data.iloc[:, spectrum_id].values
        p0 = [
            row["intensity1"], row["x_axis1"], 3,
            row["intensity2"], row["x_axis2"], 3,
            np.min(y)
        ]

        lower = [0, 375, 0.5, 0, 395, 0.5, 0]
        upper = [np.inf, 395, 15, np.inf, 420, 15, np.inf]

        try:
            popt, _ = curve_fit(
                double_gaussian, x, y,
                p0 = p0, bounds = (lower, upper), maxfev = 10000
            )
            print(f"Spectrum {spectrum_id} fitted.")

        except Exception as e:
            print(f"Spectrum {spectrum_id}: ", e)

            results.append({
                "id": spectrum_id, "mu1": 0, "mu2": 0, "fwhm1": 0, "fwhm2": 0,
                "A1": 0, "A2": 0, "area1": 0, "area2": 0, "area_ratio": 0,
                "snr": 0, "rmse": 0, "r_squared": 0, "diff": 0
            })
            continue

        A1, mu1, sigma1, A2, mu2, sigma2, C = popt
        fwhm1 = 2.35482 * sigma1
        fwhm2 = 2.35482 * sigma2

        area1 = A1 * sigma1 * np.sqrt(2*np.pi)
        area2 = A2 * sigma2 * np.sqrt(2*np.pi)
        area_ratio = area1 / area2
        
        fitted = double_gaussian(x, *popt)
        residuals = y - fitted
        rmse = np.sqrt(np.mean(residuals**2))

        ss_res = np.sum(residuals**2)
        ss_tot = np.sum((y - np.mean(y))**2)

        r_squared = 1 - ss_res / ss_tot
        noise = np.std(residuals)
        snr = np.inf if noise == 0 else np.max(y) / noise

        results.append({
            "id": spectrum_id, "mu1": mu1, "mu2": mu2, 
            "fwhm1": fwhm1, "fwhm2": fwhm2, "A1": A1, "A2": A2,
            "area1": area1, "area2": area2, "area_ratio": area_ratio,
            "snr": snr, "rmse": rmse, "r_squared": r_squared, 
            "diff": abs(mu2 - mu1)
        })

    return pd.DataFrame(results)

size = 300
file_path = (
    f"data/default LAS/"
    f"{size}x{size}/"
    f"Large Area Scan.csv"
    )
compute_time = round(0.00588271 * size ** 2 + 2.21832, 2)

minutes = int(compute_time // 60)
seconds = int(compute_time % 60)
print(f"Time to Compute: {minutes}:{seconds:02d}")

data = pd.read_csv(
    file_path,
    header = None,
    sep = "\t"
)

data = data.apply(pd.to_numeric, errors = "coerce")
data = data.dropna()

peak_summary = find_peak_locations(data)
gaussian_results = auto_gaussian_summary(data, peak_summary)

print(data[0].max(), data[0].min())
print(data.shape)
print(data.head())
print(data.iloc[:10, :5])
print(peak_summary.head())
print(peak_summary.columns)

peak_summary["diff"] = np.abs(peak_summary["x_axis1"] - peak_summary["x_axis2"])
peak_summary["ratio"] = peak_summary["intensity1"] / peak_summary["intensity2"]

peak_summary["ratio"] = np.where(
    peak_summary["ratio"] > 1,
    peak_summary["ratio"],
    1 / peak_summary["ratio"]
)

heatmap_df = peak_summary.merge(gaussian_results, on = "id")

heatmap_df["x"] = ((heatmap_df["id"] - 1) % size) + 1
heatmap_df["y"] = ((heatmap_df["id"] - 1) // size) + 1
heatmap_df["curve"] = (
    (heatmap_df["intensity1"] > 730) & (heatmap_df["intensity2"] > 730)
)

pca_data = data[(data[0] > 375) & (data[0] < 420)]
spectra_matrix = pca_data.iloc[:, 1:].T
scaler = StandardScaler()
spectra_scaled = scaler.fit_transform(spectra_matrix)

pca = PCA(n_components = 5)
scores = pca.fit_transform(spectra_scaled)
pca_scores = pd.DataFrame(scores, columns = ["PC1", "PC2", "PC3", "PC4", "PC5"])

heatmap_df = pd.concat([heatmap_df, pca_scores], axis = 1)

kmeans_vars = heatmap_df[
    [
        "mu1", "mu2", "A1", "A2", "fwhm1", "fwhm2",
        "area1", "area2", "area_ratio", "snr", "rmse", "r_squared",
        "PC1", "PC2", "PC3", "PC4", "PC5"
    ]
]

kmeans_vars = StandardScaler().fit_transform(kmeans_vars)
cluster_num = 4
kmeans = KMeans(n_clusters = cluster_num, random_state = 0, n_init = 20)

clusters = kmeans.fit_predict(kmeans_vars)
heatmap_df["cluster"] = clusters + 1

ordering = np.argsort(kmeans.cluster_centers_[:, 0])
mapping = {
    old: new + 1
    for new, old in enumerate(ordering)
}

heatmap_df["cluster"] = [
    mapping[c] * 4
    for c in heatmap_df["cluster"] - 1
]

fig = px.imshow(
    heatmap_df.pivot(index = "y", columns = "x", values = "cluster"),
    origin = "upper", aspect = "equal",
    color_continuous_scale = ["red", "yellow", "green", "blue"]
)

fig.show()

heatmap_df.to_csv(
    "../paraview_data/analysis_results.csv",
    index = False
)
